R/differential_polarity_analysis.R

```diff
@@ -95,81 +117,184 @@ RunDPA.data.frame <- function (
+      ungroup(!! sym(contrast_column)) %>%
+      mutate(ref_n = sum(!! sym(contrast_column) == reference),
+             target_n = sum(!! sym(contrast_column) %in% targets)) %>%
+      filter(ref_n > 0, target_n > 0)
+    if (nrow(test_groups) == 0) {
+      abort(glue(
+        "Found no groups with at least {min_n_obs} observations."
+      ))
+    }
+  }
+
   # Get group keys
   test_groups_keys <- test_groups %>% group_keys()
 
   if (verbose && check_global_verbosity()) {
-    cli_alert_info("Running {nrow(test_groups_keys)} tests")
+    if (length(targets) > 1) {
+      cli_alert_info(glue(
+        "Running {nrow(test_groups_keys)} tests for each ",
+        "of the following comparisons:\n"
+      ))
+    } else {
+      cli_alert_info(glue(
+        "Running {nrow(test_groups_keys)} tests for the following comparison:\n"
+      ))
+    }
+    print(glue("  - {paste(targets, reference, sep = ' vs ')}"))
   }
 
   # Split table by group keys
   test_groups <- test_groups %>% group_split()
 
-  pol_test <- lapply(test_groups, function(polarity_contrast) {
-
-    # Get numeric data values
-    x <- polarity_contrast %>% filter(.data[[contrast_column]] == target) %>% pull(morans_z)
-    y <- polarity_contrast %>% filter(.data[[contrast_column]] == reference) %>% pull(morans_z)
-
-    # Run wilcox.test
-    result <- wilcox.test(
-      x = x,
-      y = y,
-      paired = FALSE,
-      alternative = alternative,
-      conf.int = conf_int
-    )
-
-    # Tidy up results
-    result <- result %>%
-      .tidy() %>%
-      mutate(n1 = length(x), n2 = length(y), method = "Wilcoxon",
-             alternative = alternative, data_type = "morans_z",
-             target = target, reference = reference, p = signif(p.value, 3)) %>%
-      select(c("estimate", "data_type", "target", "reference", "n1", "n2", "statistic",
-               "p", "conf.low", "conf.high", "method", "alternative"))
-    return(result)
-  })
-
-  # Bind results from pol_test list
-  pol_test_bind <- do.call(bind_rows, lapply(seq_along(pol_test), function(i) {
-
-    # Add marker column
-    pol_test_with_groups <- pol_test[[i]] %>% mutate(
-      marker = test_groups_keys[i, 1, drop = TRUE]
+  # Chunk the data to enable parallel processing
+  # Decide how to split the data depending on the cl
+  if (inherits(cl, "cluster")) {
+    # Load dplyr in each cluster
+    clusterEvalQ(cl, {
+      library(dplyr)
+      library(rlang)
+      library(glue)
+      library(cli)
+    })
+    # Export variables to each cluster
+    clusterExport(cl, c(
+      "targets", "reference", "alternative", "conf_int", "polarity_metric",
+      "contrast_column",".tidy", "group_vars", "evaluate_with_catch"
+    ),
+    envir = current_env()
     )
+    # Parallel processing on cluster
+    chunks <- cut(seq_along(test_groups), length(cl))
+  } else if (is.numeric(cl)) {
+    # Parallel processing when cl is the number of child processes
+    chunks <- cut(seq_along(test_groups), cl)
```

Comment:
Here, `cl = 1` generates this error: 
```
Error in cut.default(seq_along(test_groups), cl) : 
  invalid number of intervals
```
Only for DPA though.

Reply:
Nice find! Doesn't make sense to set this to 1, but I might as well include a check for that.

Reply:
Should be fixed now!